File: LC204CountPrimes.java
package leetcode;

import java.util.Arrays;

/**
 * https://leetcode.com/problems/count-primes/
 */
public class LC204CountPrimes {
    public int countPrimes(int n) {
        if (n < 2) {
            return 0;
        }
        // Eratosthenes sieve
        boolean[] sieve = new boolean[n - 2];
        Arrays.fill(sieve, true);
        int count = 0;
        for (int i = 0; i < sieve.length; i++) {
            if (!sieve[i]) {
                continue;
            }
            count++;
            int number = 2 + i;
            for (int j = i + number; j < sieve.length; j += number) {
                sieve[j] = false;
            }
        }
        return count;
    }

    public static void main(String[] arguments) {
        System.out.println(new LC204CountPrimes().countPrimes(11));
    }
}

File: algorithms/code/leetcode/lc204_count_primes/lc204_count_primes.py
#!/usr/bin/env python3
# https://leetcode.com/problems/count-primes/

import unittest


class Solution:
    def countPrimes(self, n: int) -> int:
        if n < 2:
            return 0
        # Eratosthenes sieve
        sieve = [True] * (n - 2)
        count = 0
        for i, _ in enumerate(sieve):
            if not sieve[i]:
                continue
            count += 1
            number = 2 + i
            for j in range(i + number, len(sieve), number):
                sieve[j] = False
        return count


class TestCode(unittest.TestCase):
    def test_11(self):
        self.assertEqual(4, Solution().countPrimes(11))

    def test_1(self):
        self.assertEqual(0, Solution().countPrimes(1))
